modules/auxiliary/scanner/http/log4shell_scanner.rb

```diff
@@ -121,10 +121,14 @@ def rand_text_alpha_lower_numeric(len, bad = '')
   def run
     fail_with(Failure::BadConfig, 'The SRVHOST option must be set to a routable IP address.') if ['0.0.0.0', '::'].include?(datastore['SRVHOST'])
     @mutex = Mutex.new
+    @mutex.extend(Rex::SharedResource)
+
     @tokens = {}
+    @tokens.extend(Rex::SharedResource)
     # always disable SSL because the LDAP server doesn't use it but the setting is shared with the HTTP requests
     begin
       start_service('SSL' => false)
+      service.extend(Rex::SharedResource)
```

Comment:
If we're happy with the pattern at its surface layer, we can start to investigate adding this mixin to other core rex objects that should never be replicated - similar to initial discussions over in https://github.com/rapid7/metasploit-framework/pull/15971

Reply:
I think we have `concurrent-ruby` in the stack which provides ivars and the like for most of this stuff. Also at the point wher ewe are doing such extensions they're easier handles as runtime evals/monkey-patches on the parent class (presuming no instances of it have yet been created).
Also with mutex' not being dup'd i think you have inter-thread locking going on. My impl of this module has just the `Ref` extension wired-up as an instance_eval and it works without these other objects becoming contended across threads (when MRI finally drops the GIL, possibly today, and quite likely on non-MRI platforms with true concurrency & parallelism).

Reply:
I believe we currently rely on `concurrent-ruby` transitively, but we don't actively use those primitives in framework? Potentially this could use Ruby's queue instead of mutexes, but that's not a blocker for me since it's a standalone module - and it most likely wouldn't be the bottleneck to this module.